### ISSUE TITLE
[Question] D-PACE loss denominator after the DFlash2 integration

### ISSUE THREAD

**Issue by Lucas-TY** [sgl-project/SpecForge#828]
Hi, I am the author of D-PACE. Thank you for integrating D-PACE into SpecForge and extending it to DFlash2.

While reviewing the current implementation, I noticed that the D-PACE loss is normalized by the sum of its position weights:

```python
loss = (neg_log_q * loss_weights).sum() / loss_weights.sum()
```

D-PACE position weights are derived from expected accepted length. Dividing the weighted loss by their sum cancels the total credit mass, which may change the intended relationship between the loss and expected accepted length.

In my earlier DFlash1 experiments, this normalization appeared to affect mean accepted length, although I no longer have the logs or checkpoints needed to report the exact numbers. I have also not yet tested whether the same effect occurs with DFlash2.

Would it be possible to normalize the weighted loss by the number of valid sampled anchors instead? Would this remain compatible with the shared DFlash/DFlash2 training path?

```python
loss = weighted_loss_sum / valid_anchor_count
```

**Comment by jiapingW** [comment]
Normalizing by sum(loss_weights) removes the overall D-PACE credit mass and changes the original objective.

  Since sequences may contain different numbers of valid anchors, we propose a sequence-balanced anchor normalization:

$$
  L =
  \frac{1}{B_{\mathrm{valid}}}
  \sum_b
  \frac{N_b}{A_b}
  $$

  where (N_b) is the weighted loss numerator and (A_b) is the valid anchor count for sequence (b)，B_valid is the number of sequences containing at least one valid anchor. This is equivalent to global anchor normalization when all sequences have the same number of anchors, while preventing longer sequences from
  receiving more weight. This remains compatible with the shared DFlash/DFlash2 path and the existing chunk, gradient-accumulation, and distributed reductions.
Do you think this loss meet your expectations?

**Comment by Lucas-TY** [comment]
Yeah, I think this makes sense. Thanks for the explanation—this looks good to me and matches what I had in mind.
